clang/lib/Sema/SemaType.cpp

```diff
@@ -2340,6 +2340,12 @@ QualType Sema::BuildArrayType(QualType T, ArrayType::ArraySizeModifier ASM,
             << ArraySize->getSourceRange();
         ASM = ArrayType::Normal;
       }
+
+      // Zero length arrays are disallowed in SYCL device code.
+      if (getLangOpts().SYCLIsDevice) 
+        SYCLDiagIfDeviceCode(ArraySize->getBeginLoc(),
```

Comment:
I don't know how exactly `SYCLDiagIfDeviceCode` is working, but from it's naming I feel like it already has a check for `SYCLIsDevice` lang options.

Reply:
> I don't know how exactly `SYCLDiagIfDeviceCode` is working

You can learn it here: https://github.com/intel/llvm/blob/17aac3ccf9d5a80513bc1fd078b73fc5ecf0e0ee/clang/lib/Sema/SemaSYCL.cpp#L1444-L1457